Rcar add control of io voltage to pinctrl

xakep-amatop opened the pull request.

xakep-amatop committed:
drivers: pinctrl: pfc_rcar: add support of voltage control to pfc driver

Add support of voltage control to Renesas PFC driver. Voltage register
mappings have been added to r8a7795, r8a7796 and r8a779f SoCs.

Allow 'power-source' property for 'renesas,rcar-pfc' node. This property
will be used for configuring IO voltage on appropriate pin. For now it
is possible to have only two voltages: 1.8 and 3.3.

Note: it is possible to change voltage only for SD/MMC pins on r8a7795
      and r8a7796 SoCs, r8a779f supports more pins for configuring
      voltage, but only SD/MMC pins have been added to the array.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
drivers: rcar_mmc: add uhs pins config to renesas sd/mmc driver

Add 'power-source' property equaled to 1.8V to every eMMC pin in
order to use this config inside Renesas SD/MMC driver.

Add pin control group for UHS modes for H3ULCB, Salvator XS M3
and Spider boards to appropriate dts files. Make approprite
changes inside Renesas SD/MMC driver, e.g. use 'uhs' state for
configuring SD/MMC IO voltage as 1.8V. Both 'uhs' and 'default'
pins states have the same properties for eMMC, e.g. 1.8V.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -177,6 +177,63 @@ int pfc_rcar_set_bias(uint16_t pin, uint16_t flags)
+{
+	const struct pfc_pocctrl_reg *voltage_regs = NULL;
+
+#ifdef CONFIG_PIN_VOLTAGE_CONTROL
```
Maybe it is better to cover all logic with `#ifdef/#endif`, instead of only this line? I don't see any sense in loop below when `voltage_regs == NULL`

lorc replied:
+1

xakep-amatop replied:
done

firscity commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -189,6 +246,10 @@ int pinctrl_configure_pin(const pinctrl_soc_pin_t *pin)
 	}
 
+	if (pin->voltage != PIN_VOLTAGE_NONE) {
+		pfc_rcar_set_voltage(pin->pin, pin->voltage);
```
Should this function be int if we don't check it return value?

xakep-amatop replied:
for some of platforms we don't have implementation of `pfc_rcar_get_io_voltage_regs` and it can be one of the reason why it has to return error code, but here we haven't to do anything with the error status, but maybe somebody will need it in the future, I can change ret type from `int` to `void`

xakep-amatop replied:
done

lorc commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -177,6 +177,63 @@ int pfc_rcar_set_bias(uint16_t pin, uint16_t flags)
+	while (voltage_regs && voltage_regs->offset) {
+		for (size_t i = 0U; i < ARRAY_SIZE(voltage_regs->pins); i++) {
+			if (voltage_regs->pins[i] == pin) {
+				*bit = i;
```
why does `i` has type `size_t`, while `bit` is only `uint8_t`?

xakep-amatop replied:
copy paste from above function
I didn't pay a lot attention on types

xakep-amatop replied:
done

lorc commented on soc/arm64/renesas_rcar/gen3/pfc_r8a7795.c:
```diff
@@ -536,3 +578,7 @@ const struct pfc_drive_reg *pfc_rcar_get_drive_regs(void)
 {
 	return pfc_drive_regs;
 }
+const struct pfc_pocctrl_reg *pfc_rcar_get_io_voltage_regs(void)
```
Missing empty line before function declaration.

Did you ran `checkpatch`, BTW?

xakep-amatop replied:
checkpatch runs automatically here by GitHub
and this is the same code style as for other functions in this file

xakep-amatop replied:
done

lorc commented on boards/arm64/renesas_rcar/gen3/common_salvator_xs_h3ulcb.dts:
```diff
@@ -34,7 +34,10 @@
+	pinctrl-1 = <&sd_clk &sd_cmd &sd_ds
+			&sd_data0 &sd_data1 &sd_data2 &sd_data3
+			&sd_data4 &sd_data5 &sd_data6 &sd_data7>;
+	pinctrl-names = "default", "uhs";
```
How this supposed to work if both pinctlr states are basically the same?

xakep-amatop replied:
it is the same for mmc card, because it can work only with 1.8V
but it should be different for SD cards

lorc commented on drivers/sdhc/rcar_mmc.c:
```diff
@@ -906,7 +904,10 @@ static int rcar_mmc_change_voltage(struct sdhc_io *host_io,
+	if (!ret) {
+		host_io->signal_voltage = ios->signal_voltage;
+	}
+
```
You also supposed to reconfigure PMIC, as it should set pull-up voltage to 1.8V.

Right now you are maintaining 3.3V on pull-ups, while you already configured 1.8V in SoC... Not sure that SoC will be happy about this :)

xakep-amatop replied:
Investigating...

xakep-amatop replied:
added fixed-regulator nodes and integrate them with eMMC node and rcar-mmc driver
gpio-regulator I'll add in the future during adding SD card node to dts if you don't mind

lorc reviewed (commented):
You need to set proper pull-up voltage, before reconfiguring pinctrl or else theoretically you may damage something...

xakep-amatop committed:
drivers: regulator-fixed: extend api of driver (set/list/count voltages)

Allow to create fixed-regulator nodes without gpio for en/dis power
source. Allow properties 'regulator-min-microvolt' and
'regulator-max-microvolt' for fixed refulator. Note: they should be
equal. Add simple functions for getting list of allowed voltages, count
of voltages and dummy function for set voltage, it is used only for
checking voltage range inside regulator_set_voltage function.

The fake set voltage function inside regulator-fixed is needed only
to have common interface with other regulators and for avoiding extra
checks in place where 'regulator_set_voltage' is invoked.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop committed:
sd: request signal voltage according to host capabilities

The SD initialization function (sd_init_io) may be also used for
initialization of MMC/eMMC. eMMC/MMC may not allow 3.3 signal voltage
by default, so we should check which voltage during initialization
we may use.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

firscity commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -177,6 +177,61 @@ int pfc_rcar_set_bias(uint16_t pin, uint16_t flags)
+
+	/* Loop around all the registers to find the bit for a given pin */
+	while (voltage_regs && voltage_regs->offset) {
+		for (uint8_t i = 0U; i < ARRAY_SIZE(voltage_regs->pins); i++) {
```
`uint8_t i` should be declared outside of for loop to avoid loop initial declaration.

xakep-amatop replied:
done

firscity commented on drivers/regulator/regulator_fixed.c:
```diff
@@ -47,12 +51,49 @@ static int regulator_fixed_disable(const struct device *dev)
 static const struct regulator_driver_api regulator_fixed_api = {
 	.enable = regulator_fixed_enable,
 	.disable = regulator_fixed_disable,
+	.set_voltage = regulator_fixed_set_voltage,
```
Please add to `drivers: regulator-fixed: extend api of driver (set/list/count voltages)` more information about why we need this. I have no questions related to implementation, only about the reasons.

xakep-amatop replied:
done

firscity commented on drivers/pinctrl/pfc_rcar.c:
```diff
@@ -177,6 +177,61 @@ int pfc_rcar_set_bias(uint16_t pin, uint16_t flags)
 	return 0;
 }
 
```
mappins -> mappings in `drivers: pinctrl: pfc_rcar: add support of voltage control to pfc driver`

xakep-amatop replied:
done

lorc commented on dts/bindings/regulator/regulator-fixed.yaml:
```diff
@@ -19,7 +21,6 @@ properties:
 
   enable-gpios:
     type: phandle-array
-    required: true
```
Description above says "GPIO-controlled regulators" :)

xakep-amatop replied:
changed description

lorc commented on drivers/regulator/regulator_fixed.c:
```diff
@@ -47,12 +51,49 @@ static int regulator_fixed_disable(const struct device *dev)
+static int regulator_fixed_set_voltage(const struct device *dev,
+				       int32_t min_uv, int32_t max_uv)
+{
+	return 0;
```
You should return zero only if fixed voltage is in range `min_uv`..`max_uv`

xakep-amatop replied:
this check is performed by a caller function, pls, look at `regulator_set_voltage` common function

lorc reviewed (commented):
You are trying to mimic linux `regulator-fixed` there. I believe, this is a good approach, but some additional steps should be taken. At least you need to update YAML file description. Maybe there are other documentation parts that are needed to be updated.

Also, I can see that Linux fixed regulators use `regulator-always-on` property if they do not provide an GPIO pins. I tried to find if this is necessary, but for no avail... Could you please research this topic a bit more?

xakep-amatop commented:
> You are trying to mimic linux `regulator-fixed` there. I believe, this is a good approach, but some additional steps should be taken. At least you need to update YAML file description. Maybe there are other documentation parts that are needed to be updated.
> 
> Also, I can see that Linux fixed regulators use `regulator-always-on` property if they do not provide an GPIO pins. I tried to find if this is necessary, but for no avail... Could you please research this topic a bit more?

I've searched info and haven't found anything that can prove the next statement: "fixed regulators use `regulator-always-on` property if they do not provide an GPIO pins", but I found a lot of fixed regulators with `regulator-always-on` which have gpio(s), a few examples from Linux Kernel:
```
arch/arm64/boot/dts/rockchip/rk3399-leez-p710.dts-82-	vcc5v0_host3: vcc5v0-host3 {
arch/arm64/boot/dts/rockchip/rk3399-leez-p710.dts:83:		compatible = "regulator-fixed";
arch/arm64/boot/dts/rockchip/rk3399-leez-p710.dts-84-		regulator-name = "vcc5v0_host3";
arch/arm64/boot/dts/rockchip/rk3399-leez-p710.dts-85-		enable-active-high;
arch/arm64/boot/dts/rockchip/rk3399-leez-p710.dts-86-		gpio = <&gpio2 RK_PA2 GPIO_ACTIVE_HIGH>;
arch/arm64/boot/dts/rockchip/rk3399-leez-p710.dts-87-		pinctrl-names = "default";
arch/arm64/boot/dts/rockchip/rk3399-leez-p710.dts-88-		pinctrl-0 = <&vcc5v0_host3_en>;
arch/arm64/boot/dts/rockchip/rk3399-leez-p710.dts-89-		regulator-always-on;
arch/arm64/boot/dts/rockchip/rk3399-leez-p710.dts-90-		vin-supply = <&vcc5v0_sys>;
arch/arm64/boot/dts/rockchip/rk3399-leez-p710.dts-91-	};
```

```
arch/arm64/boot/dts/freescale/imx8dxl-evk.dts-73-	reg_fec1_io: regulator-2 {
arch/arm64/boot/dts/freescale/imx8dxl-evk.dts:74:		compatible = "regulator-fixed";
arch/arm64/boot/dts/freescale/imx8dxl-evk.dts-75-		regulator-name = "fec1_io_supply";
arch/arm64/boot/dts/freescale/imx8dxl-evk.dts-76-		regulator-min-microvolt = <1800000>;
arch/arm64/boot/dts/freescale/imx8dxl-evk.dts-77-		regulator-max-microvolt = <1800000>;
arch/arm64/boot/dts/freescale/imx8dxl-evk.dts-78-		gpio = <&max7322 0 GPIO_ACTIVE_HIGH>;
arch/arm64/boot/dts/freescale/imx8dxl-evk.dts-79-		enable-active-high;
arch/arm64/boot/dts/freescale/imx8dxl-evk.dts-80-		regulator-always-on;
```

```
arch/arm64/boot/dts/qcom/sdm630-sony-xperia-nile.dtsi-80-	cam_vana_rear_vreg: cam-vana-rear-regulator {
arch/arm64/boot/dts/qcom/sdm630-sony-xperia-nile.dtsi:81:		compatible = "regulator-fixed";
arch/arm64/boot/dts/qcom/sdm630-sony-xperia-nile.dtsi-82-		regulator-name = "cam_vana_rear_vreg";
arch/arm64/boot/dts/qcom/sdm630-sony-xperia-nile.dtsi-83-		startup-delay-us = <0>;
arch/arm64/boot/dts/qcom/sdm630-sony-xperia-nile.dtsi-84-		enable-active-high;
arch/arm64/boot/dts/qcom/sdm630-sony-xperia-nile.dtsi-85-		gpio = <&tlmm 50 GPIO_ACTIVE_HIGH>;
arch/arm64/boot/dts/qcom/sdm630-sony-xperia-nile.dtsi-86-		regulator-always-on;
```
```
arch/arm/boot/dts/tegra30-asus-transformer-common.dtsi-1594-	vdd_5v0_sys: regulator-5v {
arch/arm/boot/dts/tegra30-asus-transformer-common.dtsi:1595:		compatible = "regulator-fixed";
arch/arm/boot/dts/tegra30-asus-transformer-common.dtsi-1596-		regulator-name = "vdd_5v0_sys";
arch/arm/boot/dts/tegra30-asus-transformer-common.dtsi-1597-		regulator-min-microvolt = <5000000>;
arch/arm/boot/dts/tegra30-asus-transformer-common.dtsi-1598-		regulator-max-microvolt = <5000000>;
arch/arm/boot/dts/tegra30-asus-transformer-common.dtsi-1599-		regulator-always-on;
arch/arm/boot/dts/tegra30-asus-transformer-common.dtsi-1600-		regulator-boot-on;
arch/arm/boot/dts/tegra30-asus-transformer-common.dtsi-1601-		gpio = <&pmic 8 GPIO_ACTIVE_HIGH>;
```
```
arch/arm/boot/dts/kirkwood-dockstar.dts-62-		usb_power: regulator@1 {
arch/arm/boot/dts/kirkwood-dockstar.dts:63:			compatible = "regulator-fixed";
arch/arm/boot/dts/kirkwood-dockstar.dts-64-			reg = <1>;
arch/arm/boot/dts/kirkwood-dockstar.dts-65-			regulator-name = "USB Power";
arch/arm/boot/dts/kirkwood-dockstar.dts-66-			regulator-min-microvolt = <5000000>;
arch/arm/boot/dts/kirkwood-dockstar.dts-67-			regulator-max-microvolt = <5000000>;
arch/arm/boot/dts/kirkwood-dockstar.dts-68-			enable-active-high;
arch/arm/boot/dts/kirkwood-dockstar.dts-69-			regulator-always-on;
arch/arm/boot/dts/kirkwood-dockstar.dts-70-			regulator-boot-on;
arch/arm/boot/dts/kirkwood-dockstar.dts-71-			gpio = <&gpio0 29 0>;
```

lorc commented:
> I've searched info and haven't found anything that can prove the next statement: "fixed regulators use regulator-always-on property if they do not provide an GPIO pins"

Well, I didn't say that "they do this if and only if they do not provide GPIO pins" :) 
I just pointed that often this property is present. 


> but I found a lot of fixed regulators with regulator-always-on which have gpio(s)

Yes, this is expected. Linux automatically enables those regulators during the boot process. I am more interested in case when GPIOs are not configured. So, question is simple: if regulator can't be configured, why `always-on` property?

xakep-amatop commented:
> if regulator can't be configured, why `always-on` property?

always-on property is needed:
* to simplify enable/disable API of regulators and make the interface common
* regulators in Linux Kernel may have relationship between one another, e.g. `supply` properties, and if always-on regulator has parent, we have to enable parent even for case when the current regulator doesn't have actual usage inside any of drivers (enabled on init and building relationship)

moreover, if regulator hasn't control (e.g. gpio) and even hasn't always-on property, the core regulator logic adds always-on flag to the regulator in case if the regulator doesn't have supply

lorc reviewed (commented):
commit "drivers: rcar_mmc: add uhs pins config to renesas sd/mmc driver" does multiple things at the same time: new pinctrol options, new regulators. Could you please split it into more manageable chunks?

xakep-amatop committed:
drivers: rcar_mmc: add configuration of supply for the bus IO line power

Add nodes of fixed regulators for 1.8 and 3.3 volts to r8a779[56f]
dtsi. Use voltage from regulators for setting SD/MMC host capabilities
and configure supply for the bus IO line power.

Use regulator api to power on/off vqmmc/vmmc power sources.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>

xakep-amatop commented:
> commit "drivers: rcar_mmc: add uhs pins config to renesas sd/mmc driver" does multiple things at the same time: new pinctrol options, new regulators. Could you please split it into more manageable chunks?

done

lorc commented:
Thanks. 

`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

firscity reviewed (approved):
`Acked-by: Dmytro Firsov <dmytro_firsov@epam.com>`
